### ISSUE TITLE
Segfault in POSIX module

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6952]
Migrated from [rt.perl.org#24554](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=24554) (status was 'resolved')

Searchable as RT24554$

**Comment by p5pRT** [comment]
### From despair@cpan.org


This is a bug report for perl from despair@&#8203;cpan\.org\,
generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.




---
When calling isdigit\(undef\) function from POSIX\.pm we get segfault\.



---


---
Flags&#8203;:
&nbsp;    category=library
&nbsp;    severity=low


---
Site configuration information for perl v5\.8\.0&#8203;:

Configured by bhcompile at Tue Feb 18 22&#8203;:17&#8203;:47 EST 2003\.

Summary of my perl5 \(revision 5\.0 version 8 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.4\.20\-2\.48smp\, archname=i386\-linux\-thread\-multi
&nbsp;    uname='linux stripples\.devel\.redhat\.com 2\.4\.20\-2\.48smp \#1 smp thu feb 13 11&#8203;:44&#8203;:55 est 2003 i686 i686 i386 gnulinux '
&nbsp;    config\_args='\-des \-Doptimize=\-O2 \-march=i386 \-mcpu=i686 \-g \-Dmyhostname=localhost \-Dperladmin=root@&#8203;localhost \-Dcc=gcc \-Dcf\_by=Red Hat\, Inc\. \-Dinstallprefix=/usr \-Dprefix=/usr \-Darchname=i386\-linux \-Dvendorprefix=/usr \-Dsiteprefix=/usr \-Dotherlibdirs=/usr/lib/perl5/5\.8\.0 \-Duseshrplib \-Dusethreads \-Duseithreads \-Duselargefiles \-Dd\_dosuid \-Dd\_semctl\_semun \-Di\_db \-Ui\_ndbm \-Di\_gdbm \-Di\_shadow \-Di\_syslog \-Dman3ext=3pm \-Duseperlio \-Dinstallusrbinperl \-Ubincompat5005 \-Uversiononly \-Dpager=/usr/bin/less \-isr'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='gcc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DTHREADS\_HAVE\_PIDS \-DDEBUGGING \-fno\-strict\-aliasing \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64 \-I/usr/include/gdbm'\,
&nbsp;    optimize='\-O2 \-march=i386 \-mcpu=i686 \-g'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DTHREADS\_HAVE\_PIDS \-DDEBUGGING \-fno\-strict\-aliasing \-I/usr/local/include \-I/usr/include/gdbm'
&nbsp;    ccversion=''\, gccversion='3\.2\.2 20030213 \(Red Hat Linux 8\.0 3\.2\.2\-1\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='gcc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lgdbm \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt \-lutil
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lpthread \-lc \-lcrypt \-lutil
&nbsp;    libc=/lib/libc\-2\.3\.1\.so\, so=so\, useshrplib=true\, libperl=libperl\.so
&nbsp;    gnulibc\_version='2\.3\.1'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic \-Wl\,\-rpath\,/usr/lib/perl5/5\.8\.0/i386\-linux\-thread\-multi/CORE'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-L/usr/local/lib'

Locally applied patches&#8203;:
&nbsp;    MAINT18379



---
@&#8203;INC for perl v5\.8\.0&#8203;:
&nbsp;    /usr/lib/perl5/5\.8\.0/i386\-linux\-thread\-multi
&nbsp;    /usr/lib/perl5/5\.8\.0
&nbsp;    /usr/lib/perl5/site\_perl/5\.8\.0/i386\-linux\-thread\-multi
&nbsp;    /usr/lib/perl5/site\_perl/5\.8\.0
&nbsp;    /usr/lib/perl5/site\_perl
&nbsp;    /usr/lib/perl5/vendor\_perl/5\.8\.0/i386\-linux\-thread\-multi
&nbsp;    /usr/lib/perl5/vendor\_perl/5\.8\.0
&nbsp;    /usr/lib/perl5/vendor\_perl
&nbsp;    /usr/lib/perl5/5\.8\.0/i386\-linux\-thread\-multi
&nbsp;    /usr/lib/perl5/5\.8\.0
&nbsp;    \.



---
Environment for perl v5\.8\.0&#8203;:
&nbsp;    HOME=/home/walery
&nbsp;    LANG=ru\_RU\.koi8r
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/usr/local/bin&#8203;:/bin&#8203;:/usr/bin&#8203;:/usr/X11R6/bin&#8203;:/usr/local/tips&#8203;:/usr/java/jdk1\.3\.1/bin&#8203;:/home/walery/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @rgs

SADAHIRO Tomoyuki wrote&#8203;:
> 
> On Sat\, 29 Nov 2003 14&#8203;:00&#8203;:55 \+0100
> Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr> wrote&#8203;:
> 
> > despair@&#8203;cpan\.org \(via RT\) wrote&#8203;:
> > > 
> > > When calling isdigit\(undef\) function from POSIX\.pm we get segfault\.
> > 
> > Thanks for your report\.
> > I applied a fix to bleadperl as change \#21802\.
> > \(The bug was introduced by change \#17805\, so it's present on 5\.8\.\[12\]
> > but not 5\.8\.0\, which has another bug here\.\)
> 
> Well\, I don't know whether isalpha\(\) should be equivalent to
> /^\[\[&#8203;:alpha&#8203;:\]\]\*\\z/ or /^\[\[&#8203;:alpha&#8203;:\]\]\+\\z/\.

I actually added a test for POSIX&#8203;::isalnum\(''\) and POSIX&#8203;::isalnum\(undef\)
being true\.

> But\, for a replacement of PL\_na\, might SvPV\(\) macro be better?

The SvPV \(actually\, SvPV\_nolen\) actually occurs\, via the typemap\.

> I think another non\-string like a reference is better to be treated
> as its stringified form rather than as an empty string\.
> E\.g\. in older perl\, POSIX&#8203;::isalpha\(\[\]\) was false
> and POSIX&#8203;::isprint\(\[\]\) was true\.

Thanks for pointing this out\.
That means that I probably only need to replace the SvPOK in my patch
by a SvOK\. Agreed ?

**Comment by p5pRT** [comment]
### From @rgs

Rafael Garcia\-Suarez wrote&#8203;:
> SADAHIRO Tomoyuki wrote&#8203;:
> > Well\, I don't know whether isalpha\(\) should be equivalent to
> > /^\[\[&#8203;:alpha&#8203;:\]\]\*\\z/ or /^\[\[&#8203;:alpha&#8203;:\]\]\+\\z/\.
> 
> I actually added a test for POSIX&#8203;::isalnum\(''\) and POSIX&#8203;::isalnum\(undef\)
> being true\.
> 
> > But\, for a replacement of PL\_na\, might SvPV\(\) macro be better?
> 
> The SvPV \(actually\, SvPV\_nolen\) actually occurs\, via the typemap\.

Did I actually wrote three times actually in the same mail ? Yes\.
And moreover I replied without noticing you've supplied a patch\.
Well\, I think mine is a little bit cleaner\, macro\-proliferation\-wise
\-\- but I patched the wrong test file\.

**Comment by p5pRT** [comment]
### From BQW10602@nifty.com


On Sat\, 29 Nov 2003 14&#8203;:00&#8203;:55 \+0100
Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr> wrote&#8203;:

> despair@&#8203;cpan\.org \(via RT\) wrote&#8203;:
> > 
> > When calling isdigit\(undef\) function from POSIX\.pm we get segfault\.
> 
> Thanks for your report\.
> I applied a fix to bleadperl as change \#21802\.
> \(The bug was introduced by change \#17805\, so it's present on 5\.8\.\[12\]
> but not 5\.8\.0\, which has another bug here\.\)

Well\, I don't know whether isalpha\(\) should be equivalent to
/^\[\[&#8203;:alpha&#8203;:\]\]\*\\z/ or /^\[\[&#8203;:alpha&#8203;:\]\]\+\\z/\.
But\, for a replacement of PL\_na\, might SvPV\(\) macro be better?

I think another non\-string like a reference is better to be treated
as its stringified form rather than as an empty string\.
E\.g\. in older perl\, POSIX&#8203;::isalpha\(\[\]\) was false
and POSIX&#8203;::isprint\(\[\]\) was true\.

SADAHIRO Tomoyuki

**Comment by p5pRT** [comment]
### From BQW10602@nifty.com

<details><summary>isxxx.patch</summary>

```diff
diff -urN perl-5.8.2/ext/POSIX/POSIX.xs perl-new/ext/POSIX/POSIX.xs
--- perl-5.8.2/ext/POSIX/POSIX.xs	Wed Oct 01 02:10:22 2003
+++ perl-new/ext/POSIX/POSIX.xs	Sat Nov 29 22:12:42 2003
@@ -842,10 +842,12 @@
 
 int
 isalnum(charstring)
-	unsigned char *	charstring
+	SV *	charstring
+    PREINIT:
+	STRLEN	len;
     CODE:
-	unsigned char *s = charstring;
-	unsigned char *e = s + SvCUR(ST(0));
+	unsigned char *s = (unsigned char *) SvPV(charstring, len);
+	unsigned char *e = s + len;
 	for (RETVAL = 1; RETVAL && s < e; s++)
 	    if (!isalnum(*s))
 		RETVAL = 0;
@@ -854,10 +856,12 @@
 
 int
 isalpha(charstring)
-	unsigned char *	charstring
+	SV *	charstring
+    PREINIT:
+	STRLEN	len;
     CODE:
-	unsigned char *s = charstring;
-	unsigned char *e = s + SvCUR(ST(0));
+	unsigned char *s = (unsigned char *) SvPV(charstring, len);
+	unsigned char *e = s + len;
 	for (RETVAL = 1; RETVAL && s < e; s++)
 	    if (!isalpha(*s))
 		RETVAL = 0;
@@ -866,10 +870,12 @@
 
 int
 iscntrl(charstring)
-	unsigned char *	charstring
+	SV *	charstring
+    PREINIT:
+	STRLEN	len;
     CODE:
-	unsigned char *s = charstring;
-	unsigned char *e = s + SvCUR(ST(0));
+	unsigned char *s = (unsigned char *) SvPV(charstring, len);
+	unsigned char *e = s + len;
 	for (RETVAL = 1; RETVAL && s < e; s++)
 	    if (!iscntrl(*s))
 		RETVAL = 0;
@@ -878,10 +884,12 @@
 
 int
 isdigit(charstring)
-	unsigned char *	charstring
+	SV *	charstring
+    PREINIT:
+	STRLEN	len;
     CODE:
-	unsigned char *s = charstring;
-	unsigned char *e = s + SvCUR(ST(0));
+	unsigned char *s = (unsigned char *) SvPV(charstring, len);
+	unsigned char *e = s + len;
 	for (RETVAL = 1; RETVAL && s < e; s++)
 	    if (!isdigit(*s))
 		RETVAL = 0;
@@ -890,10 +898,12 @@
 
 int
 isgraph(charstring)
-	unsigned char *	charstring
+	SV *	charstring
+    PREINIT:
+	STRLEN	len;
     CODE:
-	unsigned char *s = charstring;
-	unsigned char *e = s + SvCUR(ST(0));
+	unsigned char *s = (unsigned char *) SvPV(charstring, len);
+	unsigned char *e = s + len;
 	for (RETVAL = 1; RETVAL && s < e; s++)
 	    if (!isgraph(*s))
 		RETVAL = 0;
@@ -902,10 +912,12 @@
 
 int
 islower(charstring)
-	unsigned char *	charstring
+	SV *	charstring
+    PREINIT:
+	STRLEN	len;
     CODE:
-	unsigned char *s = charstring;
-	unsigned char *e = s + SvCUR(ST(0));
+	unsigned char *s = (unsigned char *) SvPV(charstring, len);
+	unsigned char *e = s + len;
 	for (RETVAL = 1; RETVAL && s < e; s++)
 	    if (!islower(*s))
 		RETVAL = 0;
@@ -914,10 +926,12 @@
 
 int
 isprint(charstring)
-	unsigned char *	charstring
+	SV *	charstring
+    PREINIT:
+	STRLEN	len;
     CODE:
-	unsigned char *s = charstring;
-	unsigned char *e = s + SvCUR(ST(0));
+	unsigned char *s = (unsigned char *) SvPV(charstring, len);
+	unsigned char *e = s + len;
 	for (RETVAL = 1; RETVAL && s < e; s++)
 	    if (!isprint(*s))
 		RETVAL = 0;
@@ -926,10 +940,12 @@
 
 int
 ispunct(charstring)
-	unsigned char *	charstring
+	SV *	charstring
+    PREINIT:
+	STRLEN	len;
     CODE:
-	unsigned char *s = charstring;
-	unsigned char *e = s + SvCUR(ST(0));
+	unsigned char *s = (unsigned char *) SvPV(charstring, len);
+	unsigned char *e = s + len;
 	for (RETVAL = 1; RETVAL && s < e; s++)
 	    if (!ispunct(*s))
 		RETVAL = 0;
@@ -938,10 +954,12 @@
 
 int
 isspace(charstring)
-	unsigned char *	charstring
+	SV *	charstring
+    PREINIT:
+	STRLEN	len;
     CODE:
-	unsigned char *s = charstring;
-	unsigned char *e = s + SvCUR(ST(0));
+	unsigned char *s = (unsigned char *) SvPV(charstring, len);
+	unsigned char *e = s + len;
 	for (RETVAL = 1; RETVAL && s < e; s++)
 	    if (!isspace(*s))
 		RETVAL = 0;
@@ -950,10 +968,12 @@
 
 int
 isupper(charstring)
-	unsigned char *	charstring
+	SV *	charstring
+    PREINIT:
+	STRLEN	len;
     CODE:
-	unsigned char *s = charstring;
-	unsigned char *e = s + SvCUR(ST(0));
+	unsigned char *s = (unsigned char *) SvPV(charstring, len);
+	unsigned char *e = s + len;
 	for (RETVAL = 1; RETVAL && s < e; s++)
 	    if (!isupper(*s))
 		RETVAL = 0;
@@ -962,10 +982,12 @@
 
 int
 isxdigit(charstring)
-	unsigned char *	charstring
+	SV *	charstring
+    PREINIT:
+	STRLEN	len;
     CODE:
-	unsigned char *s = charstring;
-	unsigned char *e = s + SvCUR(ST(0));
+	unsigned char *s = (unsigned char *) SvPV(charstring, len);
+	unsigned char *e = s + len;
 	for (RETVAL = 1; RETVAL && s < e; s++)
 	    if (!isxdigit(*s))
 		RETVAL = 0;
diff -urN perl-5.8.2/ext/POSIX/t/is.t perl-new/ext/POSIX/t/is.t
--- perl-5.8.2/ext/POSIX/t/is.t	Wed Oct 01 02:10:22 2003
+++ perl-new/ext/POSIX/t/is.t	Sat Nov 29 23:00:56 2003
@@ -52,6 +52,10 @@
    " \t"       => [ qw(space) ],
 
    "abcde\001" => [],
+
+   # An empty string. Always true (al least in old days) [bug #24554]
+   ''     => [ qw(print graph alnum alpha lower upper digit xdigit
+                  punct cntrl space) ],
   );
 
 

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

despair@&#8203;cpan\.org \(via RT\) wrote&#8203;:
> 
> When calling isdigit\(undef\) function from POSIX\.pm we get segfault\.

Thanks for your report\.
I applied a fix to bleadperl as change \#21802\.
\(The bug was introduced by change \#17805\, so it's present on 5\.8\.\[12\]
but not 5\.8\.0\, which has another bug here\.\)

**Comment by p5pRT** [comment]
@rspier - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
### From @rgs

SADAHIRO Tomoyuki wrote&#8203;:
> 
> On Sat\, 29 Nov 2003 14&#8203;:00&#8203;:55 \+0100
> Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr> wrote&#8203;:
> 
> > despair@&#8203;cpan\.org \(via RT\) wrote&#8203;:
> > > 
> > > When calling isdigit\(undef\) function from POSIX\.pm we get segfault\.
> > 
> > Thanks for your report\.
> > I applied a fix to bleadperl as change \#21802\.
> > \(The bug was introduced by change \#17805\, so it's present on 5\.8\.\[12\]
> > but not 5\.8\.0\, which has another bug here\.\)
> 
> Well\, I don't know whether isalpha\(\) should be equivalent to
> /^\[\[&#8203;:alpha&#8203;:\]\]\*\\z/ or /^\[\[&#8203;:alpha&#8203;:\]\]\+\\z/\.
> But\, for a replacement of PL\_na\, might SvPV\(\) macro be better?

Thanks\, applied as \#21823\.

> I think another non\-string like a reference is better to be treated
> as its stringified form rather than as an empty string\.
> E\.g\. in older perl\, POSIX&#8203;::isalpha\(\[\]\) was false
> and POSIX&#8203;::isprint\(\[\]\) was true\.

I added also a test for the stringification of references
passed to POSIX&#8203;::isXXX\(\)\.
